### DISCUSSION TITLE
How do I version Rodauth routes?

### DISCUSSION THREAD

**Discussion by Samuelodan** [janko/rodauth-rails#231] [Q&A] [answered]
Hi there,
I want to namespace the routes that Rodauth provides, but I don't see any mention of such in the docs.
Something like `v1/create-account` would be nice.
Thanks.

**Comment by janko** [comment]
The `prefix` configuration method is for adding a path prefix to all routes, it's mentioned in [this guide](https://rodauth.jeremyevans.net/rdoc/files/doc/guides/paths_rdoc.html) and [this README section](https://github.com/jeremyevans/rodauth?tab=readme-ov-file#label-Calling+Rodauth+in+the+Routing+Tree).

```rb
prefix "/v1"
```

In rodauth-rails you don't need to make any modifications to the `route` block in the Rodauth app, it will automatically route the new prefix.

**Reply by Samuelodan** [reply]
Hmm, interesting. I previously tried both prefix and calling `r.rodauth` in the "v1" route branch, **BUT** I'm not sure that I did both of them at the same time as the guide says to do.

Thanks so much.

**Reply by Samuelodan** [reply]
Okay, so I'm away from laptop at the moment, but I want to know how to manage these versions.
My instinct is to have several config files represent the different versions, similar to if I wanted to create admin routes.

Also, in case I want admin stuff, can I nest it under the version route tree branch like so:

```rb
route do |r|
  r.on "v1" do
    r.rodauth 
    r.on "admin" do
      r.rodauth(admin)
    end
  end
end
```

And if this works, will the folder structure need to match the route tree branches like the line below?
`app/misc/v1/rodauth_main`

Oh, I think the class names would have to change for that too. Something like:
```rb
V1::RodauthApp < Rodauth::Rails::App
```

Does this make sense?

**Reply by janko** [reply, marked as answer]
Yes, essentially different versions will be processed by different Rodauth configurations (i.e. auth classes). In your example, you'd set `prefix "/v1"` in the main configuration and `prefix "/v1/admin"` in the admin configuration.

You'd probably want to version *auth classes*, just like you would version controllers. The Rodauth app should *not* be versioned, as there can be only one Rodauth app, which is called by the Rodauth middleware auto-inserted by rodauth-rails. The Rodauth app orchestrates auth classes, similarly how the Rails router orchestrates controllers.

**Reply by Samuelodan** [reply]
> The Rodauth app should not be versioned, as there can be only one Rodauth app, which is called by the Rodauth middleware auto-inserted by rodauth-rails.

Ah, I see. So it's in this one Rodauth app that'll do all the routing for the auth classes. Okay. Thanks so much. I'll let you know how it goes.

**Reply by Samuelodan** [reply]
Update: It's working now. Here's what my Rodauth app looks like:
```rb
  configure V1::RodauthMain, :v1

  # secondary configuration
  configure V1::RodauthAdmin, :admin_v1

  route do |r|
    r.rodauth(:v1)
    r.rodauth(:admin_v1)
  do
```
I realized you said there was no need to nest the route block with rodauth-rails, so I did it that way.

Interestingly, nesting it like in my earlier example, though it worked, Rails would not respond with a body in the 404 response if I sent a request to `localhost/v1/cant-login`. No routing error in the logs either. Whereas, I'd get a detailed error response if didn't hit the route tree branches like so, `localhost/vvv2/login`.

Does this mean Rodauth intercepts and rejects the requests before they get to Rails? I don't know which is better; to create new Roda route tree branches and not bother my Rails app with slightly wrong request urls, or to not do that and have Rails respond to those kinds of requests like I'd expect.

**Reply by janko** [reply]
Yes, that's exactly what happened, the Rodauth app was serving the 404 response instead of the Rails app. This has to do with how Roda's routing matchers work when used as a middleware. Normally, if you make a request that didn't hit any of Roda's routing matchers, it gets forwarded to the Rails app.

To get the routing tree to forward partially matched requests to your Rails app, you'd have to use the [pass](https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/Pass.html) Roda plugin to break out of the matched block if the request wasn't handled.

```rb
route do |r|
  r.on "v1" do
    r.rodauth 
    r.on "admin" do
      r.rodauth(admin)
      r.pass # exit for other "/v1/admin/*" requests
    end
    r.pass # exit for other "/v1/*" requests
  end
  # requests whose routing reaches the end of the route block get forwarded to the Rails app
end
```

`Rodauth::Rails::App` loads the pass plugin automatically, and [uses it](https://github.com/janko/rodauth-rails/blob/4b622c287ca2be9b0bfdc198673861af687c459e/lib/rodauth/rails/app.rb#L82-L85) when it's the one handling the prefix routing.

**Reply by Samuelodan** [reply]
Ah, it makes a lot of sense now. I've started a learning a bit about Roda in hopes that it'll help me understand a lot more about how Rodauth works. 

Thanks for all the help. At this point, I'll start working on switching my work app to Rodauth.
